[WIP] Miscellaneous fixes in ovnkube.sh

girishmg opened the pull request.

girishmg commented:
@pecameron @dcbw @shettyg PTAL

dcbw commented on dist/images/ovnkube.sh:
```diff
@@ -665,10 +665,6 @@ ovn-master () {
   wait_for_event ready_to_start_node
   echo "ovn_nbdb ${ovn_nbdb}   ovn_sbdb ${ovn_sbdb}"
 
-  # wait for northd to start
```
Hmm, why doesn't master need to wait for northd? In the containerized case we want to do so because the OVN container would be starting concurrently with the ovnkube one. Or perhaps I misunderstand.

girishmg replied:
@dcbw there is nothing in `ovnkube --master` container that depends on `ovn-northd` container being up and running. They can come up parallely or in any order.

This commit is in preparation for running our Pods with HostPID set to False. I am trying to remove all these checks where one container checks the PID of the other container when it is not required.

pecameron replied:
@girishmg At some point along the way I needed that to get ovn to come up properly. There are/were a lot of races and some of them didn't make much sense. It may no longer be needed but I would test it a lot. Maybe put a 30 sec sleep before starting it and make sure the cluster can start pods properly. If northd is required for normal operation you need make sure its there before starting normal operation. ovnkube --master is the last thing to come up and it says we are ready to go.

girishmg replied:
@pecameron good point. Will test with your suggestion and get back. Thanks.

dcbw replied:
@girishmg fair enough on northd. Does the readiness check you added cover northd then? eg we should make sure that if northd is running as a container that the entire pod is not "ready" until northd is ready and running.

girishmg replied:
@dcbw yes we have checks in place to ensure that. But, I stumbled upon an issue (something to do with onkube --node containers) that will make this change infeasible. See my comment below...

girishmg commented:
> @girishmg ovnkube.sh basically gets the cluster to come up with ovn. It does not deal with components coming and going. If a pod or container or node is down for 2 minutes (reboot time for nodes on my cluster) and then comes back, ovn must continue to operate properly based on existing conditions. (At present it does not.) There are a lot of races and not much attention to dealing with a component that goes away and comes back later.

@pecameron this is a different issue and probably should be tracked in the Project Issues page. Yes, reliability is very critical.

girishmg commented:
While testing this further I found an issue not in the ovnkube-master Pod but in ovnkube-node Pod.

I added `sleep 30` in ovn-node() before starting ovn-northd, and in ovn-master() I removed the check for ovn-northd to be up. The containers in ovnkube-master POD came up all fine.

However, in the ovnkube-node Pod the `ovn-node` container had panicked and restarted. The issue was in the following call

```
		// Create the gateway switch port in 'join' if it doesn't exist yet
		stdout, stderr, err := RunOVNNbctl("**--wait=sb**",
			"--may-exist", "lsp-add", "join", portName,
			"--", "--if-exists", "clear", "logical_switch_port", portName, "dynamic_addresses",
			"--", "lsp-set-addresses", portName, "dynamic")
```
This call timeouts since there is nothing in SB as ovn-northd is not running yet.

Currently, things work as expected because `ovnkube --node` container is blocked on `ovnkube --master` and that container is in turn blocked on `ovn-northd`
